### PR TITLE
Add SGD optimizers

### PR DESCRIPTION
**Pull request by salvacarrion**
Added support for different SGD optimizers: Vanilla SGD, Momentum, Nesterov's Accelerated Gradient, AdaGrad, RMSProp, AdaDelta, Adam and Adamax.

This optimizers allow a faster convergence rate in many problems and try to escape from saddle points, making them more robust than vanilla SGD for many cases. Moreover, visualizing its behaviour comes in handy when we're trying to understand intuition behind their basis and how suitable they are to tackle a certain problem.

### CODE REVIEW

File: orangecontrib/educational/widgets/owgradientdescent.py [hunk 1]
```diff
@@ -238,16 +250,46 @@ def __init__(self):
...
+            callback=self._opt_changed)
+
+        _r_comp = gui.doubleSpin(
+            widget=self.properties_box, master=self, value="rho", minv=1e-4,
+            maxv=1e+4, step=1e-4, label="Rho:", decimals=4,
+            alignment=Qt.AlignRight, controlWidth=90, callback=self._opt_changed)
+
+        _b1_comp = gui.doubleSpin(
+            widget=self.properties_box, master=self, value="beta1", minv=1e-5,
+            maxv=1e+5, step=1e-4, label="Beta 1:", decimals=5,
```
Comment (kernc) [inline orangecontrib/educational/widgets/owgradientdescent.py]: It's supposed to be an educational widget, yet nobody knows what these parameters mean. Perhaps add some `tooltip='...'`?

File: orangecontrib/educational/widgets/owgradientdescent.py [hunk 2]
```diff
@@ -166,6 +167,17 @@ class OWGradientDescent(OWWidget):
     auto_play_speed = settings.Setting(1)
     stochastic = settings.Setting(False)
 
+    # SGD optimizers
+    _sgd, _momentum, _nag, _adagrad, _rmsprop, _adadelta, _adam, \
+    _adamax = range(8)
```
Comment (kernc) [inline orangecontrib/educational/widgets/owgradientdescent.py]: Make this:
```py
( _OPT_SGD, _OPT_MOMENTUM, _OPT_NAG, _OPT_ADAGRAD,
 _OPT_RMSPROP, _OPT_ADADELTA, _OPT_ADAM, _OPT_ADAMAX) = range(8)
```
All-uppercase because they are constants, `_OPT` prefix to namespace them a little. Alternatively,
```py
class _Optimizer:
    SGD, MOMENTUM, NAG, ... = range(8)
```
Then use this as `self._Optimizer.SGD`. Alternatively, and **preferred**, use types (e.g. `SGD`, `Adam`, `Adamax`, ...) imported from sgd_optimizers module directly, i.e.
```py
if self.opt_type == Momentum:
```

File: orangecontrib/educational/widgets/owgradientdescent.py [hunk 3]
```diff
@@ -22,6 +22,7 @@
 from orangecontrib.educational.widgets.utils.logistic_regression \
     import LogisticRegression
 from orangecontrib.educational.widgets.utils.contour import Contour
+from orangecontrib.educational.optimizers import *
```
Comment (kernc) [inline orangecontrib/educational/widgets/owgradientdescent.py]: List imported symbols explicitly.

File: orangecontrib/educational/widgets/owgradientdescent.py [hunk 4]
```diff
@@ -297,6 +339,51 @@ def __init__(self):
...
+            elif self.opt_type == self._adadelta:
+                self.learner.sgd_optimizer = AdaDelta(self.rho)
+            elif self.opt_type == self._adam:
+                self.learner.sgd_optimizer = Adam(beta1=self.beta1,
+                                                  beta2=self.beta2)
+            elif self.opt_type == self._adamax:
+                self.learner.sgd_optimizer = Adamax(beta1=self.beta1,
+                                                    beta2=self.beta2)
+            else:
+                self.learner.sgd_optimizer = SGD()
```
Comment (kernc) [inline orangecontrib/educational/widgets/owgradientdescent.py]: With all the more exotic parameters (beta1, beta2, ...) being tunable, why `learning_rate` isn't?

File: orangecontrib/educational/widgets/utils/gradient_descent.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 from scipy.optimize import fmin_l_bfgs_b
 
 from Orange.classification import Model
-
+from orangecontrib.educational.optimizers import *
```
Comment (kernc) [inline orangecontrib/educational/widgets/utils/gradient_descent.py]: Explicit import.

File: orangecontrib/educational/widgets/utils/linear_regression.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 from orangecontrib.educational.widgets.utils.gradient_descent import \
     GradientDescent
+from orangecontrib.educational.optimizers import *
```
Comment (kernc) [inline orangecontrib/educational/widgets/utils/linear_regression.py]: Excplicit import.

File: orangecontrib/educational/widgets/utils/logistic_regression.py [hunk 7]
```diff
@@ -4,6 +4,7 @@
 
 from orangecontrib.educational.widgets.utils.gradient_descent import \
     GradientDescent
+from orangecontrib.educational.optimizers import *
```
Comment (kernc) [inline orangecontrib/educational/widgets/utils/logistic_regression.py]: Explicit import.

File: orangecontrib/educational/widgets/utils/tests/test_optimizers.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
+from orangecontrib.educational.optimizers import *
```
Comment (kernc) [inline orangecontrib/educational/widgets/utils/tests/test_optimizers.py]: Explicit.

Comment (salvacarrion) [inline orangecontrib/educational/widgets/utils/tests/test_optimizers.py]: I want to import all because I'm testing everything.

File: orangecontrib/educational/widgets/owgradientdescent.py [hunk 9]
```diff
@@ -297,6 +350,51 @@ def __init__(self):
...
+            elif self.opt_type == self._Optimizer.ADADELTA:
+                self.learner.sgd_optimizer = AdaDelta(self.rho)
+            elif self.opt_type == self._Optimizer.ADAM:
+                self.learner.sgd_optimizer = Adam(beta1=self.beta1,
+                                                  beta2=self.beta2)
+            elif self.opt_type == self._Optimizer.ADAMAX:
+                self.learner.sgd_optimizer = Adamax(beta1=self.beta1,
+                                                    beta2=self.beta2)
+            else:
+                self.learner.sgd_optimizer = SGD()
```
Comment (kernc) [inline orangecontrib/educational/widgets/owgradientdescent.py]: So still, why not parameterize the learning rate?

File: orangecontrib/educational/widgets/utils/tests/test_optimizers.py (hunk 8 above)
Comment (kernc) [inline orangecontrib/educational/widgets/utils/tests/test_optimizers.py]: Yes, you test them all. You specified them below. List all that you import here so anyone easily sees where the identifiers come from. Star imports are only suitable in package __init__s.

File: orangecontrib/educational/widgets/owgradientdescent.py (hunk 9 above)
Comment (salvacarrion) [inline orangecontrib/educational/widgets/owgradientdescent.py]: Usually, the learning rate is set when the optimizer is instantiated. But here, it is set to its corresponding value in each iteration.

Let's presume this case:
1. Select stochastic -> Vanilla SGD selected by default, with learning rate == 1.0
2. Step/Run -> Learning rate = self.learner.alpha (In this way, a user can change the learning rate at every moment, as well as the other parameters with no need to restart the process. This allow you to take an scheduled approach, e.g anneling.)

File: orangecontrib/educational/widgets/utils/tests/test_optimizers.py (hunk 8 above)
Comment (salvacarrion) [inline orangecontrib/educational/widgets/utils/tests/test_optimizers.py]:
```
from orangecontrib.educational.optimizers import SGD, Momentum, \
    NesterovMomentum, AdaGrad, RMSProp, AdaDelta, Adam, Adamax, create_opt
```

**Comment by PrimozGodec** [comment]
Closing this pull request without merging since this features are currently not required in educational k-means widget.
